gamemode/modules/commands/sv_commands.lua

```diff
@@ -0,0 +1,260 @@
+    local commandInfo = commands.list[command]
+
+    if commandInfo then
+        if not RVR.isUserGroup( ply, commandInfo.userGroup ) then
+            return "You need to be " .. RVR.getGroupName( commandInfo.userGroup ) .. " to use this command"
+        end
+
+        local newArgs, errorMsg = commands.checkArguments( commandInfo.argNames, commandInfo.argTypes, args )
+
+        if errorMsg then
+            return errorMsg, true
+        end
+
+        local msg = commandInfo.func( ply, unpack( newArgs ) )
+
+        return msg, true
+    end
+end
+
+function commands.register( names, argNames, argTypes, userGroup, func, desc )
+    if #argNames ~= #argTypes then
+        error( "There must be the same amount of argument names and types" )
+    end
+
+    if type( names ) == "string" then
+        names = { names }
+    end
+
+    for _, name in ipairs( names ) do
+        if commands.list[name] then
+            error( "A command with the same name (" .. name .. ") already exists" )
+        end
+
+        for _, argType in ipairs( argTypes ) do
+            if not commands.types[argType] then
+                error( "\"" .. argType .. "\" is not a valid argument type" )
+            end
+        end
+
+        local description = "Usage: " .. name
+
+        for i, argName in ipairs( argNames ) do
+            local argType = argTypes[i]
+
+            description = description .. " " .. argName .. ":" .. argType
+        end
+
+        if desc then
+            description = description .. "\nDescription: " .. desc
+        end
+
+        commands.list[name] = {
+            argNames = argNames,
+            argTypes = argTypes,
+            userGroup = userGroup,
+            func = func,
+            description = description
+        }
+    end
+end
+
+local function onPlayerSay( ply, text )
+    if string.sub( text, 1, 1 ) ~= "!" then return end
+
+    text = string.sub( text, 2 )
+
+    local args = processArguments( text )
+    local command = table.remove( args, 1 )
+
+    local msg, validCommand = processCommand( ply, command, args )
+
+    if msg then
+        ply:ChatPrint( msg )
+    end
+
+    if not validCommand then
+        ply:ChatPrint( "Command \"" .. command .. "\" does not exist" )
+    else
+        return ""
+    end
+end
+
+hook.Add( "PlayerSay", "RVR_Commands_onPlayerSay", onPlayerSay )
+
+local function onRunConsoleCommand( len, ply )
+    local argsStr = net.ReadString()
+
+    local args = processArguments( argsStr )
+    local command = table.remove( args, 1 )
+
+    local msg, validCommand = processCommand( ply, command, args )
+
+    if msg then
+        ply:PrintMessage( HUD_PRINTCONSOLE, msg )
+    end
+
+    if not validCommand then
+        ply:PrintMessage( HUD_PRINTCONSOLE, "Command \"" .. command .. "\" does not exist" )
+    end
+end
+
+net.Receive( "RVR_Commands_runConsoleCommand", onRunConsoleCommand )
+
+commands.register( "help", {"command"}, {"string"}, RVR_USER_ALL, function( command )
```

Comment:
```suggestion
commands.register( "help", { "command" }, { "string" }, RVR_USER_ALL, function( command )
```

Reply:
I was told not to put spaces in these for lua styling?

Reply:
By who? >:O

Reply:
You have code elsewhere with spaces around `{}`
```lua
    if type( names ) == "string" then
        names = { names }
    end
```

Reply:
I'm a fan of spaces inside for parenthesis and brackets. Just looks a bit nicer

Reply:
Phatso told me that @samuelWilliams99

Reply:
Time to beat up @brandonsturgeon then >:(